[AMD] Rewrite canonicalize pointers to use 1:N conversion

makslevental opened the pull request:
### TL;DR (too long, didn't review)

This PR re-enables the `tritonamdgpu-canonicalize-pointers` pass[^1]. The PR is effectively a complete rewrite of the original pass, which walked the AST and mutated IR in-place, using the new [`1:N` dialect conversion framework](https://github.com/llvm/llvm-project/pull/116470). Recall a "fat pointer" is a tuple-like `(%baseptr, %offsetptr)` - the current (original) pass keeps this tuple in a global data structure while the new/rewritten pass emits this tuple into the IR as an `unrealized_cast(%baseptr, %offsetptr)`[^2].

Note, this PR also rewrites the existing lit test (see [this comment below](https://github.com/triton-lang/triton/pull/5329#issuecomment-2552629207)).

### Pass outline

The pass structure/action is roughly:

1. Perform an approximate sparse dataflow analysis to find all transitive uses for `tt.func` args that are `tt.ptr`s; legalize only these ops;
2. Rewrite all operations' `use`s and `result`s to be `(%baseptr, %offsetptr)` using `ConversionPattern`s that takes the new `OneToNOpAdaptor`, which automatically forwards both `%baseptr` and `%offsetptr` through `adaptor.getOperands()`[^3];
3. Clean up remaining `unrealized_casts` (currently only handling one category of such remaining casts but can be extended to handle all; see bullet 1 in TODOs).

### Some pre-emptive call outs

Right up front I'll say this took a long time to figure out because **a)** the conversion framework is hugely complex **b)** it's being currently rewritten to be more robust/stable. As a consequence, the implementation is complex but I've tried hard to **a)** simplify as much as possible **b)** comment/note subtleties **c)** put in ample `assert`s and checks to clarify intent and gracefully fail. So some things to call out:

1. I called the dataflow analysis approximate because it does not actually use [DataFlow/SparseAnalysis](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Analysis/DataFlow/SparseAnalysis.cpp) and instead computes a forward slice using the heuristic "transfer function" that users of an op with a `tt.ptr` operand should be rewritten. This heuristic works because the forward slice starts from `tt.ptr` args on a `tt.func` and ends at `tt.store`, which has no results. Note, there's no reason why this component of the pass can't be a true `SparseAnalysis` implementation, it's just that this rewrite has already taken way longer than I expected (so I leave that for a possible follow-up).
2. The pass uses no global `TypeConverter` but uses local `TypeConverter`s, in `BranchInterface`/`RegionInterface` patterns. This is because **a)** we are not actually converting operand/result types (we are converting number of operands/results) **b)** the conversion framework expects/handles this lack of a `TypeConverter` exactly [the way we want](https://github.com/llvm/llvm-project/blob/399c3a78a2577c6fc68bba7f301901a0e66e87ed/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1179-L1185). The local type converters are used for ops that couple to basic blocks (`^bb`s) that need to have their signatures rewritten (i.e., the ops for which we need to do `rewriter.applySignatureConversion(block, *conversion, &localTypeConverter)`). That's `scf.for`, `scf.while`, `cf.br` and `cf.cond_br` (not needed for `scf.if` which has no `bb` args).
3. `tt.func` is handled differently from all of the other ops - it is not rewritten at all. Instead, for every `%arg: tt.ptr` arg, we insert into the new body `%c0 = arith.constant 0 : i32` and `%newarg = unrealized_cast(%arg, %c0) : tt.ptr` (manually, not done by the conversion framework) and replace all uses of `%arg` by `%newarg`. These are then unpacked to `(%arg, %c0)` using `replaceOpWithMultiple` so that they "magically" appear in `adaptor.getOperands()`. Then at the end, currently, these are the only unreconciled casts (because they are the only ones **not** inserted by the conversion framework) and we materialize them by just replacing uses of `%newarg` with `%arg`. 
4. `scf.if` needs to be handled specially; since it has no operands but can `yield` results, we need to rewrite it only after its `yield`s have been rewritten. This is not straightforward because the dialect conversion [does a preorder walk](https://github.com/llvm/llvm-project/blob/6ab8401f53deddbd79f930ba2bec2f824c9567e3/mlir/lib/Transforms/Utils/DialectConversion.cpp#L2705). To work around this we define legality for `scf.if` to be dependent on whether its `yield`s have been rewritten (using two `UnitAttr`s on those `yield`s). Thus, `scf.if` is "legal" and not rewritten until after the results of the `yield`s are known. 

### TODOs (decisions to be made)

~~1. The new pass only reconciles the intial `unrealized_cast`s that come from `tt.func` args (see bullet 3 above). That means ops that aren't currently handled (e.g., `tt.bitcast`, `tt.scan`, etc.) will fail to be legalized and the pass will fail (e.g., [test_inductor_cummax_bool](https://github.com/triton-lang/triton/blob/main/python/test/regression/test_functional_regressions.py#L261-L263)). The old pass would just "materialize the fat pointer" for unknown operations i.e., fallback to default handling. I can certainly restore this behavior but I would prefer not to until we've done more testing that the pass is correct (i.e., that's why I've kept it behind `AMDGCN_USE_BUFFER_OPS`).~~
~~2. The current pass propagates pointer attributes (such as `tt.pointer_range`) via attributes on ops. This leads to accumulating lots of attributes (such as `arith.OverflowFlags` and etc.) in the [global data structure](https://github.com/triton-lang/triton/blob/f12eae03adf741ab253a2fed64c08878f24ecc07/third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp#L105). I would prefer not to do this and given that we are currently discussing alternative approaches to handling such compiler hints, I did not carry over the same behavior. Again, I can certainly restore it but I propose we first decide how we would like to handle compiler hints and then afterward possibly restore it (if we decide that these attributes are how we want to implement hinting).~~

### Trivial TODOs 

Still need to add patterns for `triton::AtomicCASOp`, `triton::AtomicRMWOp`, `triton::PtrToIntOp`.

### Shoutouts

Thanks @giuseros for all the advice and explanations. Shoutout @matthias-springer for both rewriting the conversion framework and advising me on how it actually works.


cc @danzimm @SamGinzburg 


[^1]: I haven't actually moved it out of the flag but it's now usable with the `AMDGCN_USE_BUFFER_OPS` flag whereas it wasn't prior.
[^2]: In reality it's the conversion framework that materializes this tuple as `unrealized_cast(%baseptr, %offsetptr)` and then reconciles/DCEs all the casts automatically.
[^3]: The `unrealized_cast`s are completely "transparent" to the patterns, see [`ConversionPatternRewriterImpl::remapValues`](https://github.com/llvm/llvm-project/blob/399c3a78a2577c6fc68bba7f301901a0e66e87ed/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1161).

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -611,28 +611,71 @@ LogicalResult PointerCanonicalizer::rewriteForOp(scf::ForOp forOp,
+      operandNum + 1 - forOp.getNumControlOperands();
+
+  SmallVector<Value> newInitArgs(
+      forOp.getInitArgs().take_front(numLeadingInitsIncludingCurOp));
```
I would remove the currentOp. It was there before exactly because we were pushing back the new operands. We can now get rid of it (and make the IR way simpler).

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -611,28 +611,71 @@ LogicalResult PointerCanonicalizer::rewriteForOp(scf::ForOp forOp,
+
+    newForOp.getBody()->insertArgument(numLeadingInitsIncludingCurOp + 1,
+                                       offset.getType(), offset.getLoc());
+    newForOp.getBody()->insertArgument(numLeadingInitsIncludingCurOp + 1,
```
Shouldn't be `numLeadingInitsIncludingCurOp + 2`

makslevental replied:
No because you insert offset and then base behind it

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -611,28 +611,71 @@ LogicalResult PointerCanonicalizer::rewriteForOp(scf::ForOp forOp,
   rewriteOpMap[forOp] = newForOp;
 
-  newForOp->setOperand(operandNum, tensorPtr);
+  {
```
Basically, if you remove the current operand from the the loop arguments, you won't need this anymore. This will be removed by canonicalization anyway, but it's here only because we have to pass "something" as the current operand.

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -611,28 +611,71 @@ LogicalResult PointerCanonicalizer::rewriteForOp(scf::ForOp forOp,
+  }
+
   OpOperand *forOperand = &newForOp->getOpOperand(operandNum);
   // This is making sure we propagate the visit from the forOp result
```
Stale comment tbr

giuseros commented on test/TritonGPU/amd/amd-canonicalize-pointers.mlir:
As a warmup comment, could you add a test to check the issue we were having? Something on the line of:
```
for (%ptrA,%ptrB){
   yield %ptrB, %ptrA
}
```

makslevental commented:
@giuseros @antiagainst 

There's a problem with the checked in lit tests (...lots of problems).

The main issue is that they're not stable wrt `-cse -canonicalize`. 

What do I mean? If I run both my version of the pass and the old version of the pass with `-cse -canonicalize`, then my pass produces the same output as the old pass (naturally I need to do this because my pass inserts ops in different orders than the old pass).

The problem is that if I do `-cse -canonicalize` then the old pass no longer passes the lit tests because tons of things get canonicalized and DCEd away.

E.g., [these](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L569) [tests](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L624) [that](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L489) [have](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L434) no returns

And these tests with control flow but nothing happens in the [bodies](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L694-L701) [so the args gets forward](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L328-L335) all the way through.

And these [whiles that don't](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L330) [update the condition](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L696).

Also [lots](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L397-L398) [of](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L539C5-L539C10) [tests](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L630C5-L630C13) [with](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L657-L658) [unused](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L686C5-L688) [`arith.constant`](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L353C6-L354) [SSAs](https://github.com/triton-lang/triton/blob/main/test/TritonGPU/amd/amd-canonicalize-pointers.mlir#L714-L717).

In summary - I think I've rewritten the pass correctly authentically but the the lit tests aren't helping me actually verify that. So I will need to rewrite them from scratch too I think...

makslevental commented:
On the bright side, the regression is fixed:

```
Launching pytest with arguments test_functional_regressions.py::test_iv_dependent_matmul --no-header --no-summary -q in /home/mlevental/dev_projects/triton/python/test/regression

============================= test session starts ==============================
collecting ... collected 5 items

test_functional_regressions.py::test_iv_dependent_matmul[pre_load] 
test_functional_regressions.py::test_iv_dependent_matmul[post_load] 
test_functional_regressions.py::test_iv_dependent_matmul[post_pre_mixed] 
test_functional_regressions.py::test_iv_dependent_matmul[post_load_two_iters] 
test_functional_regressions.py::test_iv_dependent_matmul[post_load_three_iters] 

======================== 5 passed, 1 warning in 26.87s =========================
PASSED [ 20%]PASSED [ 40%]PASSED [ 60%]PASSED [ 80%]PASSED [100%]
```

giuseros commented:
> The main issue is that they're not stable wrt -cse -canonicalize.
Yes, I see what you mean and I agree, you will probably need to rewrite a lot of them. This has always been an issue with the canonicalization pass

> On the bright side, the regression is fixed:
This is amazing! Thanks @makslevental !

makslevental commented:
@antiagainst @giuseros I have updated the tests using `-canonicalize` and [mlir/utils/generate-test-checks.py](https://github.com/llvm/llvm-project/blob/5f096fd2216001296b809002ee474ee6d7f06e0e/mlir/utils/generate-test-checks.py). PTAL and see if there's still reasonable. @giuseros lucky for me there are already two tests that test two pointers in `scf.for` (`forOp` and `forOp2`).

makslevental commented:
Also FYI this is blocked until we catch up to https://github.com/llvm/llvm-project/pull/116470

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+    SmallVector<ValueRange> packedRets;
+    for (unsigned i = 0, offset = 0; i < valRangeLens.size(); i++) {
+      // skip %cond
+      if (remappedInits[i].size() == 1 &&
```
Why this would skip only `%cond`?

makslevental replied:
because we don't "remap" `%cond` ever (we never do `1:N` for that value)

giuseros replied:
What is not clear to me is why we cannot handle this as we handle other non-ptr parameters. I mean I think you are referring to this:
```
    %6 = scf.while (%arg1 = %5, %arg2 = %cond) : (tensor<1024x!tt.ptr<f32>, #blocked>, i1) -> (tensor<1024x!tt.ptr<f32>, #blocked>) {
        scf.condition(%arg2) %arg1 : tensor<1024x!tt.ptr<f32>, #blocked>
    } do {
        ^bb0(%arg1: tensor<1024x!tt.ptr<f32>, #blocked>):
            scf.yield %arg1, %cond : tensor<1024x!tt.ptr<f32>, #blocked>, i1
    }
```
Why `cond` should be handled differently from other non-pointer parameters? `%5` is our pointer and it will be split in `{baseptr, offset}`, but the rest will be untouched (I think). Sorry in advance if the question if naive :)

makslevental replied:
Sorry I don't understand the question;

>  `but the rest will be untouched` 

This `continue` is exactly how they are untouched (because they are skipped). Anything that doesn't have a `1:2` conversion (as we receive in `OpOneToNAdaptor` is skipped because the only expected conversions are `1:2` (so any other conversion is not expected and in fact should never be written in this pass). The extra check here `remappedInits[i].getType()[0].isInteger(1))` should actually be an `assert` because we expect the *only* other things in the `scf.while` is the condition.

makslevental replied:
Sorry I was doing a bit of "peephole" commenting. The reason we skip cond here is because here we're packing the results and `scf.while` doesn't return/result the cond variable.

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+      };
+    };
+
+    auto propagateCanNarrowToBlock = [this](Block *block,
```
Could this be a separate function? I see you are defining this in multiple places, and it looks similar

makslevental replied:
They're all subtly different depending on the op and they're so small that trying to refactor into shared functionality would make them all much more complicated.

makslevental replied:
Went ahead and refactored this for `cf.br`, `cf.cond_br` and `scf.while` but left `scf.for` which is "special".

giuseros commented on test/TritonGPU/amd/amd-canonicalize-pointers.mlir:
Tests are fine. But I was wondering if you could add a unit test that exposes the problematic IR for which we disabled this pass previously (basically, a `forOp` that swaps the `yield` arguments). In-advance apology if you already added the test and I missed it :)

makslevental replied:
Added

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+    return;
+  }
+
+  // TODO(max): this doesn't make sense - ops have all sorts of dialect
```
Was this to be removed?

antiagainst replied:
It's fine to keep some TODO here but we should remove the commented out code--they are disconnected and confusing to be here.

giuseros commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -278,77 +143,41 @@ Value getScalarConstant(IRRewriter &rewriter, Location loc, Value expr) {
 // Narrowing logic
 // For now we allow to narrow down to 32 bits only in the following case:
 // - `baseOffset` is 32-bits and `addOffset`(64-bits) is zero
+// TODO(max): is this correct?
```
I think the comment is wrong (and should be rewritten), and the code is a bit convoluted and probably deserves at least a comment. The first time we use a non-uniform offset (and only the first time), we need to splat the zero offset and then add the non-uniform offset. I am catching the splat here .

giuseros reviewed (commented):
Thanks for the marvelous job @makslevental ! It all looks fine! Only few nits and some answers to your `TODO`s.

makslevental commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+      // absent then the `scf.if` is legal. Once both `yields` have been
+      // rewritten (the corresponding attributes have been added), we report the
+      // `scf.if` as illegal, and it will be rewritten (the pattern will fire).
+      return !(ifOp->hasAttr(kSCFThenRewrittenAttr) and
```
whoops (but aren't we all surprised that `and` works lol)

makslevental commented:
FYI: blocked on https://github.com/llvm/llvm-project/pull/123122

matthias-springer commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,979 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+            rewriter.create<arith::ConstantIntOp>(funcOp.getLoc(), 0, bitness);
+        auto dummyCast = rewriter.create<UnrealizedConversionCastOp>(
+            arg.getLoc(), TypeRange{arg.getType()}, ValueRange{arg});
+        rewriter.replaceUsesOfBlockArgument(arg, dummyCast.getResult(0));
```
The semantics of `replaceUsesOfBlockArgument` are not well defined. Actually, I was trying to remove it for a while... Personally, I try to avoid it because I always worry that something may break when using it.

Not saying that your code here is incorrect. But given that `replaceUsesOfBlockArgument` is somewhat fragile, I'd recommend to rewrite this pattern as follow:

- Create a new `tt.func` op.
- Inline the body of the old `tt.func` op.
- I think you won't need `replaceUsesOfBlockArgument`.
- This may also work around the issue with the circular mapping because you are replacing with a block argument of a new block.
- Erase the old `tt.func` op.

Would that work?

makslevental replied:
So the thing is, the explicit goal here is to not change the signature of the `tt.func` (because that's the kernel API right). So this ~hack/workaround was to create initial "tuple" objects `(base, offset=0)` for the args to the `tt.func`:

```mlir
tt.func @kernel(%A : !tt.ptr<f32>) {
  ... // users of %A
}
```

into

```mlir
tt.func @alloc(%A : !tt.ptr<f32>) {
  %cst_0 = arith.constant
  %noop_cast = unrealized_cast(%A) : !tt.ptr<f32>
  ... // users %A become users of %noo_cast
}
```

and then let `1:N` dialect conversion do its thing remapping `%noop_cast` with just `(%A, %cst_0)`. 

This now (maybe prior as well?) induces a circular mapping that causes the infinte while from the PR I closed. But here I can refactor this same pattern into just 

```mlir
tt.func @alloc(%A : !tt.ptr<f32>) {
  %cst_0 = arith.constant
  %first_fat_ptr = unrealized_cast(%A, %cst_0) : !tt.ptr<f32>
}
```

and then special case check for `%first_fat_ptr`. The problem is there are other semantics changes in `1:N` conversion (from the most recent PR that you landed) that prevent that from being a solution as well (hence the other opened and [closed PR](https://github.com/llvm/llvm-project/pull/123154)).

matthias-springer replied:
> So the thing is, the explicit goal here is to not change the signature of the tt.func 

I'm not sure if this actually works, but I was thinking of creating a new func op even though the signature does not change. Sounds useless at first, but the mapping is no longer circular, because you map from the old bbarg to the new bbarg + an extra SSA value.

makslevental replied:
Oh that is an interesting idea. Let me try that!

makslevental replied:
This indeed breaks the circular mapping (thanks) but replacing the funcop means all the nested ops are now [skipped because they've also been "replaced"](https://github.com/llvm/llvm-project/blob/0a3ef6a4d69602ea0c4913e7bd751ba7e2fa95e9/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1571). But working around this still leaves me the problems from [https://github.com/llvm/llvm-project/pull/123154], namely that the semantics of `lookupOrDefault` have changed to no longer return the "most recent" remapping.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -442,16 +268,118 @@ PointerCanonicalizer::decomposeOffsetFromExpr(Location loc, Value expr,
+
+struct FatPointers {
+  struct FatPtrAttrs {
+    FatPtrAttrs(const FatPtrAttrs &other) = default;
+    FatPtrAttrs &operator=(const FatPtrAttrs &other) = default;
+    // for map default insert
+    FatPtrAttrs() = default;
+    bool canNarrow = false;
+    llvm::SmallDenseMap<StringRef, Attribute> attributes;
+    friend bool operator==(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return lhs.canNarrow == rhs.canNarrow && lhs.attributes == rhs.attributes;
+    }
+    friend bool operator!=(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return !(lhs == rhs);
+    }
```
Style: Can we reorganize it a bit so that we have constructors first, and then operators, and then data fields. Use some empty lines to separate them.

Similarly for the outer class. A bit better for readability.

makslevental replied:
As we discussed during 1-1, I might actually remove this thing entirely (depending on whether `int-range-narrow` works for us). But yea sure I can re-org here.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -442,16 +268,118 @@ PointerCanonicalizer::decomposeOffsetFromExpr(Location loc, Value expr,
+static const std::string kSCFIfOpYieldFatPtrOffsets =
+    kPtrCanonPrefix + "scf-if-yield-fatptr-offsets__";
+
+struct FatPointers {
```
Can you add some comments regarding what this struct is for and how it is used throughout given it's a very important piece of data structure.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+    // other to indicate to the parent IfOp that the result type can now be
+    // rewritten and not before.
+    if (auto ifOp = dyn_cast<scf::IfOp>(yieldOp->getParentOp())) {
+      if (ifOp.thenBlock() == yieldOp->getBlock()) {
+        rewriter.modifyOpInPlace(ifOp, [&] {
+          ifOp->setDiscardableAttr(kSCFThenRewrittenAttr,
+                                   rewriter.getUnitAttr());
+        });
+      } else {
+        rewriter.modifyOpInPlace(ifOp, [&] {
+          ifOp->setDiscardableAttr(kSCFElseRewrittenAttr,
+                                   rewriter.getUnitAttr());
+        });
```
Nit: Maybe do something like

```c++
StringRef branchAttr = ifOp.thenBlock() == yieldOp->getBlock() ?
  kSCFThenRewrittenAttr : kSCFElseRewrittenAttr;
rewriter.modifyOpInPlace(...);
```

makslevental replied:
lol duh whoops

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+      auto dummyCast = rewriter.create<UnrealizedConversionCastOp>(
+          arg.getLoc(), TypeRange{arg.getType()}, ValueRange{arg, zeroOffset});
+      rewriter.replaceAllUsesExcept(arg, dummyCast.getResult(0), dummyCast);
+      // TODO(max): why is this true?
```
This is effectively the initial state for `canNarrow`? @giuseros

makslevental replied:
Yea @giuseros already responded to most of these TODO questions above

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+                   ConversionPatternRewriter &rewriter) const override {
+    // Exhaustive checking we're converting ONLY unrealized_casts inserted (by
+    // the 1:N conversion) in ConvertFuncOp.
+    assert(std::distance(castOp->getUses().begin(), castOp->getUses().end()) >
```
`castOp.use_empty()`?

antiagainst reviewed (changes requested):
Cool! Thanks for rewriting this--using the 1:N infra is much cleaner and more robust. Overall lgtm; I just have a few nits here and there.

makslevental commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+  walkAndApplyPatterns(getOperation(), std::move(patterns));
+  patterns.clear();
+
+  auto func = getFunc(getOperation());
```
I typically put decl/init as close to use as possible so that if a use gets removed then so does the decl/init but sure moving this one higher makes sense.

makslevental commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+static FailureOr<tt::FuncOp> getFunc(Operation *moduleOp) {
+  SmallVector<tt::FuncOp> funcOps;
+  moduleOp->walk<WalkOrder::PreOrder>([&funcOps](tt::FuncOp op) {
+    if (op.isPrivate())
```
I had this here because in one of tests I needed a simple way to prevent DCE - so I had a private `tt.func dummy(...)` and then I did `call @dummy(...)`. But I replaced that with just an unregistered dialect `"dummy"(...)`. So cool I can remove this and replace with the size == 1 check.

makslevental commented on third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp:
```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+  target.addDynamicallyLegalDialect<arith::ArithDialect>(isLegal);
+
+  // Rewrite the rest of the ops.
+  // Note we *do not* declare unrealized_cast an
+  // illegal op here in order that the whole conversion passes, even if there
+  // are tt ops that we do not currently support (their operands will be handled
+  // by ConvertUnimplementedOpUnrealizedCasts below). Note we *do* add
+  // ConvertFuncOpArgsUnrealizedCasts because that is necessary for
+  // "initializing" the chain of fat pointers starting from tt.func tt.ptr args.
```
ultra-nit lol

makslevental commented:
@antiagainst I left the "design decision" about propagation as a `TODO` because I need to think about it some more and no point in blocking on that "thinking" (i.e., I restored the original behavior).

antiagainst reviewed (changes requested):
This largely lgtm now; thanks for addressing all the comments! One last request: can you temporarily enable the pass in `compiler.py` so we can run runtime tests for it? Right now it's only lit tested. Once that's okay we can turn it off by default again before landing.

makslevental commented:
> This largely lgtm now; thanks for addressing all the comments! One last request: can you temporarily enable the pass in `compiler.py` so we can run runtime tests for it? Right now it's only lit tested. Once that's okay we can turn it off by default again before landing.

Done
